File: grooves-api/src/main/groovy/com/github/rahulsom/grooves/api/snapshots/Snapshot.java
package com.github.rahulsom.grooves.api.snapshots;

import com.github.rahulsom.grooves.api.AggregateType;
import com.github.rahulsom.grooves.api.events.BaseEvent;

/**
 * Marks a class as a snapshot. This supports both temporal and versioned access.
 *
 * @param <AggregateT>  The Aggregate this snapshot works over
 * @param <SnapshotIdT> The type for the snapshot's {@link #getId()} field
 * @param <EventIdT>    The type for the {@link EventT}'s id field
 * @param <EventT>      The base type for events that apply to {@link AggregateT}
 *                
 * @author Rahul Somasunderam
 */
public interface Snapshot<
        AggregateT extends AggregateType, SnapshotIdT, EventIdT,
        EventT extends BaseEvent<AggregateT, EventIdT, EventT>>
        extends
        VersionedSnapshot<AggregateT, SnapshotIdT, EventIdT, EventT>,
        TemporalSnapshot<AggregateT, SnapshotIdT, EventIdT, EventT> {

    @Override
    default void setLastEvent(EventT aidBaseEvent) {
        this.setLastEventTimestamp(aidBaseEvent.getTimestamp());
        this.setLastEventPosition(aidBaseEvent.getPosition());
    }

}

File: grooves-types/src/main/kotlin/com/github/rahulsom/grooves/api/snapshots/Snapshot.kt
package com.github.rahulsom.grooves.api.snapshots

import com.github.rahulsom.grooves.api.events.BaseEvent
import com.github.rahulsom.grooves.api.snapshots.internal.BaseSnapshot

/**
 * Marks a class as a snapshot. This supports both temporal and versioned access.
 *
 * @param [AggregateT] The Aggregate this snapshot works over
 * @param [SnapshotIdT] The type for [BaseSnapshot.id]
 * @param [EventIdT] The type for [BaseEvent.id]
 * @param [EventT] The base type for events that apply to [AggregateT]
 *
 * @author Rahul Somasunderam
 */
interface Snapshot<
    AggregateT,
    SnapshotIdT,
    EventIdT,
    in EventT : BaseEvent<AggregateT, EventIdT, in EventT>,
    > :
    VersionedSnapshot<AggregateT, SnapshotIdT, EventIdT, EventT>,
    TemporalSnapshot<AggregateT, SnapshotIdT, EventIdT, EventT>